homeassistant/components/kdeconnect/config_flow.py

```diff
@@ -0,0 +1,153 @@
+    CONNECT_TIMEOUT,
+    DATA_KEY_CLIENT,
+    DATA_KEY_NAMES,
+    DOMAIN,
+)
+from .helpers import ensure_running
+
+_LOGGER = logging.getLogger(__name__)
+
+
+async def try_pair(hass: HomeAssistant, device_id: str) -> dict[str, Any]:
+    """Try pairing to the selected device."""
+
+    client = cast(KdeConnectClient, hass.data[DOMAIN][DATA_KEY_CLIENT])
+
+    device = client.connected_devices[device_id]
+    result = await device.pair()
+    if result == PairingResult.REJECTED:
+        raise PairingRejected()
+
+    if result == PairingResult.TIMEOUT:
+        raise PairingTimeout()
+
+    if result == PairingResult.ACCEPTED:
+        cert = device.certificate
+        assert cert is not None
+
+        return {
+            CONF_DEVICE_NAME: device.device_name,
+            CONF_DEVICE_ID: device_id,
+            CONF_DEVICE_TYPE: device.device_type.value,
+            CONF_DEVICE_CERT: cert.public_bytes(serialization.Encoding.PEM).decode(
+                "utf-8"
+            ),
+            CONF_DEVICE_INCOMING_CAPS: list(device.incoming_capabilities),
+            CONF_DEVICE_OUTGOING_CAPS: list(device.outgoing_capabilities),
+        }
+
+    assert False  # pragma: no cover
+
+
+class ConfigFlow(config_entries.ConfigFlow, domain=DOMAIN):
+    """Handle a config flow for KDEConnect."""
+
+    VERSION = 1
+
+    async def async_step_user(
+        self, user_input: dict[str, Any] | None = None
+    ) -> FlowResult:
+        """Handle the initial step."""
+        await ensure_running(self.hass)
+
+        client = cast(KdeConnectClient, self.hass.data[DOMAIN][DATA_KEY_CLIENT])
+
+        if user_input is None:
+            return self.async_show_device_selection(client)
+
+        if user_input[CONF_DEVICE_NAME] == CONF_REFRESH:
+            await client.advertise_once()
+            # Allow some time for devices to respond to our request
+            await asyncio.sleep(CONNECT_TIMEOUT)
+            return self.async_show_device_selection(client)
+
+        device_id = self.hass.data[DOMAIN][DATA_KEY_NAMES][user_input[CONF_DEVICE_NAME]]
+
+        return await self.async_step_pair(device_id)
+
+    async def async_step_pair(self, device_id: str) -> FlowResult:
+        """Handle the pairing after selecting a device."""
+        await self.async_set_unique_id(device_id)
+        self._abort_if_unique_id_configured()
+
+        client = self.hass.data[DOMAIN][DATA_KEY_CLIENT]
+
+        errors = {}
+        try:
+            data = await try_pair(self.hass, device_id)
+        except PairingRejected:
+            errors["base"] = "rejected"
+        except PairingTimeout:
+            errors["base"] = "timeout"
+        except Exception:  # pylint: disable=broad-except
+            _LOGGER.exception("Unexpected exception")
+            errors["base"] = "unknown"
+        else:
+            return self.async_create_entry(title=data[CONF_DEVICE_NAME], data=data)
+
+        return self.async_show_device_selection(client, errors=errors)
```

Comment:
Typically you can only show a form on a step that posts back to the same step. So since `async_show_device_selection` posts to step `user` -- so i think this really implies that the pair step is really just part of the user step. Can you inline most of this in the user step?

Reply:
I think if this is restructured as a single step then the need to store devices in hass.data is removed (e.g. am aybe  helper function just returns the devices rather than querying devices and showing the form all together like it does today)